runtime/tools/python/ttrt/common/callback.py

```diff
@@ -0,0 +1,440 @@
+
+        for index, row in enumerate(reader):
+            if index == 2:
+                add_key(
+                    dram_memory_usage,
+                    l1_memory_usage,
+                    "L1",
+                    "largest_contiguous_free_block (bytes) : per bank",
+                    row[1].strip(),
+                )
+
+
+def parse_memory_csv_files(
+    detailed_memory_usage_file_path,
+    memory_usage_summary_file_path,
+    l1_usage_summary_file_path,
+):
+    dram_memory_usage = {}
+    l1_memory_usage = {}
+
+    parse_detailed_memory_usage_file(
+        dram_memory_usage, l1_memory_usage, detailed_memory_usage_file_path
+    )
+    parse_memory_usage_summary_file(
+        dram_memory_usage, l1_memory_usage, memory_usage_summary_file_path
+    )
+    parse_l1_usage_summary_file(
+        dram_memory_usage, l1_memory_usage, l1_usage_summary_file_path
+    )
+
+    return dram_memory_usage, l1_memory_usage
+
+
+def memory(callback_runtime_config, binary, program_context, op_context):
+    import ttrt.runtime
+    import ttrt.binary
+
+    device = callback_runtime_config.device
+    logging = callback_runtime_config.logging
+    logging.debug("executing memory dump")
+    loc = ttrt.runtime.get_op_loc_info(op_context)
+    debug_str = ttrt.runtime.get_op_debug_str(op_context)
+
+    device.dump_memory_report()
+    memory_dump_dir_path = f"{get_ttrt_metal_home_path()}/generated/reports"
+
+    # read generated memory reports and store in condensed memory_report
+    dram_memory_usage, l1_memory_usage = parse_memory_csv_files(
+        f"{memory_dump_dir_path}/detailed_memory_usage.csv",
+        f"{memory_dump_dir_path}/memory_usage_summary.csv",
+        f"{memory_dump_dir_path}/l1_usage_summary.csv",
+    )
+
+    op_memory_report = {}
+    op_memory_report["loc"] = loc
+    op_memory_report["debug_str"] = debug_str
+    op_memory_report["dram"] = dram_memory_usage
+    op_memory_report["l1"] = l1_memory_usage
+    callback_runtime_config.memory_report[
+        callback_runtime_config.callback_counter()
+    ] = op_memory_report
+
+
+"""
+-----------------------DEBUGGER CALLBACK-----------------------
+"""
+
+
+def debugger(callback_runtime_config, binary, program_context, op_context):
+    import pdb
+
+    device = callback_runtime_config.device
+    logging = callback_runtime_config.logging
+    logging.debug("starting pdb debugger")
+    pdb.set_trace()
+
+
+def callback(callback_runtime_config, binary, program_context, op_context):
+
+    if callback_runtime_config.enable_golden:
+        golden(callback_runtime_config, binary, program_context, op_context)
+
+    if callback_runtime_config.enable_memory:
+        memory(callback_runtime_config, binary, program_context, op_context)
+
+    if callback_runtime_config.enable_debugger:
+        debugger(callback_runtime_config, binary, program_context, op_context)
+
+
+def get_callback_fn(callback_runtime_config):
+    return partial(callback, callback_runtime_config)
```

Comment:
I'll leave this refactor up to you, but just my 2 cents:
Each of these callbacks might be better organized inside of their own subdirectories & files, `callbacks/golden.py`, `callbacks/memory.py`, `callbacks/debugger.py`.  They could also each have their own `Config` object.

You could compose them together with something like:
```python
def get_callback_fn(golden_config=None, memory_config=None, debugger_config):
    callbacks = []
    if golden_config is not None:
        callbacks.append(partial(golden, golden_config)
    if memory_config is not None:
        callbacks.append(partial(memory, memory_config)
    if debugger_config is not None:
        callbacks.append(partial(debugger, debugger_config)

    def callback_fn(*args, **kwargs):
        for fn in callbacks:
            fn(*args, **kwargs)

    return callback_fn
```